src/Lean/Message.lean

```diff
@@ -64,7 +64,10 @@ inductive MessageData where
   |  compose           : MessageData → MessageData → MessageData
   /-- Tagged sections. `Name` should be viewed as a "kind", and is used by `MessageData` inspector functions.
     Example: an inspector that tries to find "definitional equality failures" may look for the tag "DefEqFailure". -/
-  | tagged            : Name → MessageData → MessageData
+  | tagged             : Name → MessageData → MessageData
+  /-- A lazy message data. This can be used to delay possibly costly computations of `MessageData`
+    that may not be used in the end, such as a `throwError` in a tactic inside `try`.  -/
+  | thunk              : Thunk MessageData → MessageData
```

Comment:
Hey @nomeata! I'm still confused as to why this needs a new constructor given that `ofPPFormat` exists. The issue was that using `ofPPFormat` does not properly generate highlightable spans. Could that not be fixed without adding a new constructor?

Reply:
Worth exploring. Can you define `++` on `FormatWithInfos`?

Reply:
What's the difference between implementing this as a new constructor and defining `MessageData.thunk x` as `.ofPPFormat fun _ => x.get`?

Reply:
I am assuming that if the new constructor isn't needed then neither is `BaseIO.ofThunk`, so that all concerns about purity/evaluation order go away.

Reply:
Because `x` can return `MessageData`, not just `FormatWithInfo`, so you can use all the message data related functionality – embedding goals with `ofGoal`, different contexts for different subcomponents, etc.

In a way, `.ofPPFormat` is a constructor that bundles three orthogonal features that could all be their own constructor (ignoring positivity requirements):

 * It allows lazy creation of a message (with a `hasSyntheticSorry` peeking)
 * It gives access to an implicit context, so that you compose MessageData in pure code without having to say `(← ppTerm e)` everywhere
 * It extends `.ofFormat` with the ability to add Info annotations
 
One might argue that `.ofPPFormat` could be split into three constructors, one for each of these features, then it would be easier to put them together in more flexible ways.

I may well be missing some obvious solution here, though, I don’t think I have completely penetrated this part of the code base.

Reply:
Hmm, reading the code some more it seems that renumbering the tags in `FormatWithInfos` to be able to concatenate two of them, and thus life the `Format` API to `FormatWithInfos`, may be possible as well. Doesn’t feel particularly great to have yet another Format-API-wrapper (besides `Format` itself and `MessageData`).

Reply:
> I am assuming that if the new constructor isn't needed then neither is BaseIO.ofThunk, so that all concerns about purity/evaluation order go away.

Yes, that would be good indeed, it’s a bit of a cannon and sparrows situation. We might for example define a `IOThunk` variant where `.get` is monadic, to fix the purity worries.

Reply:
Isn't `IOThunk` just spelled `IO`?

Reply:
I'd expect an `IOThunk` to internally also store the result and run the action at most once, something we want for lazy MessageData.